Add docs for ADK integration for deployment (#4092)
langchain-ai/docs: src/langsmith/deploy-google-adk.mdx

---
title: Deploy Google ADK agents
sidebarTitle: Deploy Google ADK
description: Deploy Google Agent Development Kit (ADK) agents to LangSmith Agent Server using the deployments-wrap-sdk package.
---

This guide shows you how to deploy a [Google Agent Development Kit (ADK)](https://github.com/google/adk-python) agent on [LangSmith Agent Server](/langsmith/agent-server) using the [`deployments-wrap-sdk`](https://pypi.org/project/deployments-wrap-sdk/) package.

`deployments-wrap-sdk` provides a thin wrapper that turns a configured ADK `Runner` into a LangGraph-compatible graph, so you can deploy ADK agents without writing the [Functional API](/oss/langgraph/functional-api) glue yourself. The wrapper:

> - Bridges ADK sessions to Agent Server's [checkpoint persistence](/langsmith/agent-server#persistence), so session state survives restarts and resumes across runs.
> - Forwards ADK token events through LangGraph's streaming pipeline, so partial tokens show up in [`stream_mode="messages"`](/langsmith/streaming) and in [LangGraph Studio](/langsmith/studio).
> - Automatically enables [LangSmith tracing](/langsmith/trace-with-google-adk) for ADK when `LANGSMITH_TRACING` is set.

Reviewer: ```suggestion
- Bridges ADK sessions to Agent Server's [checkpoint persistence](/langsmith/agent-server#persistence), so session state survives restarts and resumes across runs.
- Forwards ADK token events through LangGraph's streaming pipeline, so partial tokens show up in [`stream_mode="messages"`](/langsmith/streaming) and in [LangSmith Studio](/langsmith/studio).
- Automatically enables [LangSmith tracing](/langsmith/trace-with-google-adk) for ADK when `LANGSMITH_TRACING` is set.
```

I think LangGraph should be LangSmith in these bullet points. That is, "LangSmith Deployment's streaming pipeline"

## Prerequisites

- Python 3.11+

> - LangGraph CLI for local dev and deployment

Reviewer: ```suggestion
- [LangGraph CLI](/langsmith/cli)for local dev and deployment
```

> - A LangSmith API key — see [Create an account and API key](https://docs.langchain.com/langsmith/create-account-api-key)

Reviewer: ```suggestion
- A LangSmith API key, refer to [Create an account and API key](https://docs.langchain.com/langsmith/create-account-api-key)
```

> - A Google AI API key if you use Gemini models — see [Google AI Studio](https://aistudio.google.com/api-keys)

Reviewer: ```suggestion
- A Google AI API key if you use Gemini models, refer to [Google AI Studio](https://aistudio.google.com/api-keys)
```

## Installation

> Install the package with the `google-adk` extra. The extra pulls in `google-adk` itself and other dependencies needed for the wrapper:

Reviewer: ```suggestion
Install the package with the `google-adk` extra. The extra pulls in `google-adk` and other dependencies needed for the wrapper:
```

```bash
pip install "deployments-wrap-sdk[google-adk]"
```

<Note>
The PyPI distribution name is `deployments-wrap-sdk`, but the Python import path is `saf_sdk`. Both refer to the same package.
</Note>

## Quickstart

> This minimal example builds an "echo" agent that does not require any model API key — it short-circuits the LLM call so you can verify the deployment plumbing end-to-end before wiring in a real model.

Reviewer: ```suggestion
This minimal example builds an agent that returns the input as its response and does not require a model API key. The agent bypasses the LLM call so you can verify that the deployment works correctly before connecting a real model.
```
Adjusting the language to make it less idiomatic.

Create `agent.py`:

```python agent.py
from google.adk.agents import Agent
from google.adk.models.llm_response import LlmResponse
from google.adk.runners import Runner
from google.genai.types import Content, Part
from saf_sdk.adk import LangsmithSessionService, wrap

def echo_callback(callback_context, llm_request):
    """Return the user's message instead of calling a real model."""
    user_text = ""
    if callback_context.user_content and callback_context.user_content.parts:
        for part in callback_context.user_content.parts:
            if part.text:
                user_text += part.text
    return LlmResponse(
        content=Content(role="model", parts=[Part(text=f"echo: {user_text}")])
    )

agent = wrap(
    Runner(
        agent=Agent(
            name="echo_agent",
            model="gemini-2.0-flash",
            instruction="Echo the user message.",
            before_model_callback=echo_callback,
        ),
        app_name="adk_echo",
        session_service=LangsmithSessionService(),
    )
)
```

Two things are essential:

> 1. **Pass `LangsmithSessionService()`** as the runner's `session_service`. `wrap()` raises a `TypeError` if you forget — Agent Server needs this hook to load and save ADK session state through its checkpointer.

Reviewer: ```suggestion
1. **Pass `LangsmithSessionService()`** as the runner's `session_service`. `wrap()` raises a `TypeError` if you forget. Agent Server needs this hook to load and save ADK session state through its checkpointer.
```

2. **Export the wrapped `agent`** as a module-level variable. Agent Server imports this symbol when serving the graph.

> For a real agent, drop the `before_model_callback` and configure a model normally — for example, Gemini via `model="gemini-2.0-flash"` with `GOOGLE_API_KEY` set, or Claude/OpenAI via ADK's LiteLLM adapter (`google.adk.models.lite_llm.LiteLlm`, available through `google-adk[extensions]`).

Reviewer: ```suggestion
For a real agent, drop the `before_model_callback` and configure a model directly. For example, use Gemini by setting `model="gemini-2.0-flash"` with `GOOGLE_API_KEY` set, or use Claude/OpenAI via ADK's LiteLLM adapter (`google.adk.models.lite_llm.LiteLlm`, available through `google-adk[extensions]`).
```

## Project layout

A deployable project needs three files:

```
my-adk-agent/
├── agent.py              # exports the wrapped agent
├── langgraph.json        # Agent Server config
└── pyproject.toml        # Python dependencies
```

> `langgraph.json` points Agent Server at the exported symbol:

Reviewer: ```suggestion
[`langgraph.json`](/langsmith/application-structure#configuration-file-concepts) points Agent Server at the exported symbol:
```

```json langgraph.json
{
  "$schema": "https://langgra.ph/schema.json",
  "dependencies": ["."],
  "graphs": {
    "adk_echo": "./agent.py:agent"
  },
  "env": ".env"
}
```

`pyproject.toml` declares dependencies:

```toml pyproject.toml
[project]
name = "my-adk-agent"
version = "0.0.1"
requires-python = ">=3.11"
dependencies = [
    "deployments-wrap-sdk[google-adk]>=0.0.1",
]
```

## Install dependencies

```bash
pip install -e .
```

## Run locally

Start the local Agent Server with the [LangGraph CLI](/langsmith/cli):

```bash
langgraph dev
```

> This serves the agent at `http://127.0.0.1:2024` and opens [LangGraph Studio](/langsmith/studio) so you can chat with the agent. Send a request directly with `curl`:

Reviewer: ```suggestion
This serves the agent at `http://127.0.0.1:2024` and opens [LangSmith Studio](/langsmith/studio) so you can chat with the agent. Send a request directly with `curl`:
```

```bash
# Create a thread
THREAD=$(curl -s -X POST http://127.0.0.1:2024/threads \
  -H "Content-Type: application/json" -d '{}' | python -c "import sys, json; print(json.load(sys.stdin)['thread_id'])")

# Run the agent and wait for the final response
curl -s -X POST "http://127.0.0.1:2024/threads/$THREAD/runs/wait" \
  -H "Content-Type: application/json" \
  -d '{
    "assistant_id": "adk_echo",
    "input": {"messages": [{"type": "human", "content": "Hello"}]}
  }'
```

## Deploy to LangSmith

Once the agent runs locally, deploy it to LangSmith with `langgraph deploy`:

```bash
langgraph deploy --name my-adk-agent
```

> See [Deploy to cloud](/langsmith/deploy-to-cloud) for environment configuration, deployment types, and revision management. For self-hosted setups, see [Self-hosted deployments](/langsmith/self-hosted).

Reviewer: ```suggestion
For environment configuration, deployment types, and revision management, refer to [Deploy to cloud](/langsmith/deploy-to-cloud). For self-hosted setups, refer to [Self-hosted deployments](/langsmith/self-hosted).
```

## Enable tracing

`wrap()` calls `langsmith.integrations.google_adk.configure_google_adk()` automatically whenever LangSmith tracing is enabled, so all you need to do is set the environment variables on the deployment:

```bash .env
LANGSMITH_API_KEY=your-langsmith-api-key
LANGSMITH_TRACING=true
LANGSMITH_PROJECT=my-adk-agent     # optional
GOOGLE_API_KEY=your-google-api-key
```

> Traces show agent invocations, tool calls, and LLM interactions in the [LangSmith UI](/langsmith/observability). For more on the underlying tracing integration, see [Trace Google ADK applications](/langsmith/trace-with-google-adk).

Reviewer: ```suggestion
[Traces](/langsmith/observability) show agent invocations, tool calls, and LLM interactions in the [LangSmith UI](https://smith.langchain.com). For more on the underlying tracing integration, see [Trace Google ADK applications](/langsmith/trace-with-google-adk).
```

## API reference

### `wrap(runner)`

Wraps a configured `google.adk.runners.Runner` and returns a LangGraph `Pregel` graph that can be exported from your module and served by Agent Server.

| Argument | Type | Description |
|---|---|---|
| `runner` | `google.adk.runners.Runner` | A configured ADK Runner. Its `session_service` **must** be a `LangsmithSessionService`. |

**Returns:** A `Pregel` graph whose name is `runner.app_name`.

**Raises:** `TypeError` if `runner.session_service` is not a `LangsmithSessionService`.

If `runner.agent` defines an `output_key`, that key's value is also exposed on the graph's output, in addition to `messages`. This is what makes ADK structured-output agents (`output_schema=...`, `output_key=...`) work with Studio and the `/runs/wait` response.

### `LangsmithSessionService`

> A `google.adk.sessions.BaseSessionService` implementation backed by Agent Server's checkpoint store. The wrapper manages session lifecycle automatically — it creates a session on the first turn of a thread, loads it from the checkpoint on subsequent turns, and writes the updated session back when the run completes.

Reviewer: ```suggestion
A `google.adk.sessions.BaseSessionService` implementation backed by Agent Server's checkpoint store. The wrapper manages session lifecycle automatically. It creates a session on the first turn of a thread, loads it from the checkpoint on subsequent turns, and writes the updated session back when the run completes.
```

Use a fresh instance per `Runner`:

```python
session_service = LangsmithSessionService()
```

You should not need to call its methods directly; `wrap()` drives them through ADK's normal session lifecycle.

### `ADKInput`

The default input schema for a wrapped agent.

| Field | Type | Description |
|---|---|---|

> | `messages` | `list[AnyMessage]` | Required. Conversation messages; the wrapper sends `messages[-1].content` to the ADK runner as the new user message. |

Reviewer: ```suggestion
| `messages` | `list[AnyMessage]` | (Required) Conversation messages; the wrapper sends `messages[-1].content` to the ADK runner as the new user message. |
```

> | `state_delta` | `dict[str, Any] \| None` | Optional. Passed through to `runner.run_async(state_delta=...)` to mutate ADK session state for this turn. |

Reviewer: ```suggestion
| `state_delta` | `dict[str, Any] \| None` | (Optional) Passed through to `runner.run_async(state_delta=...)` to mutate ADK session state for this turn. |
```
